ntoskrnl/ex/time.c

```diff
@@ -225,85 +319,23 @@ BOOLEAN
+    if (!result)
     {
-        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        DPRINT1("ExpGetTimeZoneId failed\n");
         return FALSE;
     }
+    DPRINT("ExpTimeZoneId is %lu\n", ExpTimeZoneId);
 
-    /* Get the default bias */
-    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
-
-    if (ExpTimeZoneInfo.StandardDate.Month != 0 &&
-        ExpTimeZoneInfo.DaylightDate.Month != 0)
-    {
-        /* Get this years standard start time */
-        if (!RtlCutoverTimeToSystemTime(&ExpTimeZoneInfo.StandardDate,
-                                        &StandardTime,
-                                        CurrentBootTime,
-                                        TRUE))
-        {
-            DPRINT1("RtlCutoverTimeToSystemTime() for StandardDate failed!\n");
-            return FALSE;
-        }
-
-        /* Get this years daylight start time */
-        if (!RtlCutoverTimeToSystemTime(&ExpTimeZoneInfo.DaylightDate,
-                                        &DaylightTime,
-                                        CurrentBootTime,
-                                        TRUE))
-        {
-            DPRINT1("RtlCutoverTimeToSystemTime() for DaylightDate failed!\n");
-            return FALSE;
-        }
-
-        /* Determine the time zone id and update the time zone bias */
-        if (DaylightTime.QuadPart < StandardTime.QuadPart)
-        {
-            if ((CurrentBootTime->QuadPart >= DaylightTime.QuadPart) &&
-                (CurrentBootTime->QuadPart < StandardTime.QuadPart))
-            {
-                DPRINT("Daylight time!\n");
-                ExpTimeZoneId = TIME_ZONE_ID_DAYLIGHT;
-                ExpTimeZoneBias.QuadPart += (LONGLONG)ExpTimeZoneInfo.DaylightBias * TICKSPERMINUTE;
-            }
-            else
-            {
-                DPRINT("Standard time!\n");
-                ExpTimeZoneId = TIME_ZONE_ID_STANDARD;
-                ExpTimeZoneBias.QuadPart += (LONGLONG)ExpTimeZoneInfo.StandardBias * TICKSPERMINUTE;
-            }
-        }
-        else
-        {
-            if ((CurrentBootTime->QuadPart >= StandardTime.QuadPart) &&
-                (CurrentBootTime->QuadPart < DaylightTime.QuadPart))
-            {
-                DPRINT("Standard time!\n");
-                ExpTimeZoneId = TIME_ZONE_ID_STANDARD;
-                ExpTimeZoneBias.QuadPart += (LONGLONG)ExpTimeZoneInfo.StandardBias * TICKSPERMINUTE;
-            }
-            else
-            {
-                DPRINT("Daylight time!\n");
-                ExpTimeZoneId = TIME_ZONE_ID_DAYLIGHT;
-                ExpTimeZoneBias.QuadPart += (LONGLONG)ExpTimeZoneInfo.DaylightBias * TICKSPERMINUTE;
-            }
-        }
-    }
-    else
-    {
-        ExpTimeZoneId = TIME_ZONE_ID_UNKNOWN;
-    }
+    ExpTimeZoneBias = NewTimeZoneBias;
 
-    /* Change it for user-mode applications */
+    /* Change SharedUserData->TimeZoneBias for user-mode applications */
```

Comment:
(Since I cannot comment at the correct place of code, I'm doing it there):
@Doug-Lyons , could you also replace the `/* Convert system time from local time to UTC */` do-while loop below, that retrieves `SharedUserData->SystemTime` into `CurrentTime` below (old lines 316-320 / new lines 348-352) with a `KeQuerySystemTime(&CurrentTime);` too?

In addition I think that the resetting of the `SharedUserData->SystemTime` after that loop should be changed so that the ordering of setting the High1/2Time and LowPart should respect the ordering as done also here:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/clock.c;hb=8e799e2d3eeaf350acda68700577671b405580d9#l54

Reply:
@HBelusca, I have not been able to figure out how to change this section of code to use ```KeQuerySystemTime(&CurrentTime);```. To make this match the replaced code, it seems that for the parameter I need to use ```&CurrentTime.u```, but this produces a compile time error that the parameter needs to be a ```PLARGE_INTEGER```. This variable type is complex and uses ```union``` and ```struct``` together and both of these are foreign to me.
If you will give me complete line-by-line code changes, then I will try this again. Sorry, but this is above me for now. Thanks.

Reply:
I have changed the ordering to High2Time, LowPart, and then High1Time in two places. Please confirm that this is what you wanted. Thanks.

Reply:
> @HBelusca, I have not been able to figure out how to change this section of code to use `KeQuerySystemTime(&CurrentTime);`. To make this match the replaced code, it seems that for the parameter I need to use `&CurrentTime.u`, [...]

No, since [`KeQuerySystemTime()` takes a pointer to `LARGE_INTEGER` for input/output](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/clock.c;hb=8e799e2d3eeaf350acda68700577671b405580d9#l184), and since `CurrentTime` is already a `LARGE_INTEGER`, then the suggested `KeQuerySystemTime(&CurrentTime);` should compile OK and works as expected.

Reply:
> I have changed the ordering to High2Time, LowPart, and then High1Time in two places. Please confirm that this is what you wanted. Thanks.

I confirm your changes are OK.

Reply:
OK. now with the ```KeQuerySystemTime(&CurrentTime);```. Thanks.